### PR TITLE
Set entire state of an account

### PR DESCRIPTION
**Pull request by miiu96**
## Reasoning behind the pull request
- set the entire state on an account 
  

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: node/chainSimulator/components/testOnlyProcessingNode.go [hunk 1]
```diff
@@ -382,26 +385,40 @@ func (node *testOnlyProcessingNode) collectClosableComponents(apiInterface APICo
...
 
-	userAccount, ok := account.(state.UserAccountHandler)
-	if !ok {
-		return errors.New("cannot cast AccountHandler to UserAccountHandler")
+	_, err = accountsAdapter.Commit()
+	if err != nil {
+		return err
 	}
 
+	return nil
```
Comment (sstanculeanu) [inline node/chainSimulator/components/testOnlyProcessingNode.go:411]:
```suggestion
	return err
```

**Review by sstanculeanu — APPROVED** [review]

File: node/chainSimulator/chainSimulator.go [hunk 2]
```diff
@@ -202,7 +203,34 @@ func (s *simulator) SetState(address string, state map[string]string) error {
...
+	for _, state := range stateSlice {
+		addressBytes, err := addressConverter.Decode(state.Address)
+		if err != nil {
+			return err
+		}
+
+		shardID := sharding.ComputeShardID(addressBytes, s.numOfShards)
+		if bytes.Equal(addressBytes, core.SystemAccountAddress) {
+			// for system account address use the provided shard ID
+			shardID = state.ShardID
```
Comment (iulianpascalau) [inline node/chainSimulator/chainSimulator.go:224]: I think that, for the system account we should use all 3 shards.

File: node/chainSimulator/components/testOnlyProcessingNode.go [hunk 3]
```diff
@@ -382,26 +385,40 @@ func (node *testOnlyProcessingNode) collectClosableComponents(apiInterface APICo
...
+	err = accountsAdapter.SaveAccount(userAccount)
 	if err != nil {
 		return err
 	}
 
-	userAccount, ok := account.(state.UserAccountHandler)
-	if !ok {
-		return errors.New("cannot cast AccountHandler to UserAccountHandler")
+	_, err = accountsAdapter.Commit()
+	if err != nil {
```
Comment (iulianpascalau) [inline node/chainSimulator/components/testOnlyProcessingNode.go:407]: yes, as @sstanculeanu suggested, we need to return the err here and remove this if case

File: node/chainSimulator/components/testOnlyProcessingNode.go [hunk 4]
```diff
@@ -382,26 +385,40 @@ func (node *testOnlyProcessingNode) collectClosableComponents(apiInterface APICo
 	}
 }
 
-// SetState will set the provided state for the given address
-func (node *testOnlyProcessingNode) SetState(address []byte, keyValueMap map[string]string) error {
+// SetKeyValueForAddress will set the provided state for the given address
+func (node *testOnlyProcessingNode) SetKeyValueForAddress(address []byte, keyValueMap map[string]string) error {
+	userAccount, err := node.getUserAccount(address)
```
Comment (iulianpascalau) [inline node/chainSimulator/components/testOnlyProcessingNode.go:390]: what happens if the address is not in the shard maintained by this testOnlyProcessingNode instance?

File: node/chainSimulator/components/testOnlyProcessingNode.go [hunk 5]
```diff
@@ -411,7 +428,48 @@ func (node *testOnlyProcessingNode) SetState(address []byte, keyValueMap map[str
 		}
 	}
 
-	err = accountsAdapter.SaveAccount(account)
+	return nil
+}
+
+// SetStateForAddress will set the state for the give address
+func (node *testOnlyProcessingNode) SetStateForAddress(address []byte, addressState *dtos.AddressState) error {
+	userAccount, err := node.getUserAccount(address)
```
Comment (iulianpascalau) [inline node/chainSimulator/components/testOnlyProcessingNode.go:436]: what happens if the address is not in the shard maintained by this testOnlyProcessingNode instance?

File: node/chainSimulator/components/testOnlyProcessingNode.go [hunk 6]
```diff
@@ -411,7 +428,48 @@ func (node *testOnlyProcessingNode) SetState(address []byte, keyValueMap map[str
...
+	if err != nil {
+		return err
+	}
+
+	// set nonce to zero
+	userAccount.IncreaseNonce(-userAccount.GetNonce())
```
Comment (iulianpascalau) [inline node/chainSimulator/components/testOnlyProcessingNode.go:442]: :bulb:

File: node/chainSimulator/components/testOnlyProcessingNode.go [hunk 7]
```diff
@@ -411,7 +428,48 @@ func (node *testOnlyProcessingNode) SetState(address []byte, keyValueMap map[str
...
+	}
+
+	rootHash, err := base64.StdEncoding.DecodeString(addressState.RootHash)
+	if err != nil {
+		return err
+	}
+	userAccount.SetRootHash(rootHash)
+
+	accountsAdapter := node.StateComponentsHolder.AccountsAdapter()
+	err = accountsAdapter.SaveAccount(userAccount)
```
Comment (iulianpascalau) [inline node/chainSimulator/components/testOnlyProcessingNode.go:472]: return err directly on L478?

File: node/chainSimulator/dtos/state.go [hunk 8]
```diff
@@ -0,0 +1,17 @@
+package dtos
+
+// AddressState will hold the address state
+type AddressState struct {
+	Address string `json:"address"`
+	// ShardID: This field is needed for the system account address (it is the same on all shards).
+	ShardID          uint32            `json:"shardID,omitempty"`
```
Comment (iulianpascalau) [inline node/chainSimulator/dtos/state.go:7]: I think we do not need this. The data on all SystemAccounts (on all shards) should be the same.

**Review by iulianpascalau — CHANGES_REQUESTED** [review]

File: node/chainSimulator/components/testOnlyProcessingNode.go (hunk 1 above)
Comment (miiu96) [reply node/chainSimulator/components/testOnlyProcessingNode.go:411]: done

File: node/chainSimulator/components/testOnlyProcessingNode.go (hunk 3 above)
Comment (miiu96) [reply node/chainSimulator/components/testOnlyProcessingNode.go:407]: done

File: node/chainSimulator/components/testOnlyProcessingNode.go (hunk 7 above)
Comment (miiu96) [reply node/chainSimulator/components/testOnlyProcessingNode.go:472]: done

File: node/chainSimulator/components/testOnlyProcessingNode.go (hunk 5 above)
Comment (miiu96) [reply node/chainSimulator/components/testOnlyProcessingNode.go:436]: same here

File: node/chainSimulator/dtos/state.go (hunk 8 above)
Comment (miiu96) [reply node/chainSimulator/dtos/state.go:7]: removed

File: node/chainSimulator/chainSimulator.go (hunk 2 above)
Comment (miiu96) [reply node/chainSimulator/chainSimulator.go:224]: removed

**Review by iulianpascalau — APPROVED** [review]

**Review by sstanculeanu — APPROVED** [review]

**Pull request merged**